### PR TITLE
RSP opcode fixes

### PR DESCRIPTION
**Pull request by GhostlyDark**
Ported from:
https://gitlab.com/parallel-launcher/parallel-n64/-/commit/4f7b8a6081f7d02bac26e041f1baf5d5e91875f2
https://gitlab.com/parallel-launcher/parallel-n64/-/commit/3052f79c162a232301fdf725038f813c1ec02edb
https://gitlab.com/parallel-launcher/parallel-n64/-/commit/f379d606e4dd000b72ae90273a1b1874b51f5307
https://gitlab.com/parallel-launcher/parallel-n64/-/commit/07ce4bfd7528cde68939ad2ce1a06c2c7029ec15#06a9a5c035040370b6d9d811526d9586179b3176_52_52

Reference:
https://github.com/libretro/parallel-rsp/pull/10

### CODE REVIEW

**Comment by GhostlyDark** [comment]
@mpharoah Anything missing here?

**Comment by mpharoah** [comment]
Looks good

**Review by mpharoah — APPROVED** [review]

File: src/device/rcp/rsp/rsp_core.c [hunk 1]
```diff
@@ -137,68 +145,68 @@ static void fifo_pop(struct rsp_core* sp)
...
-    if (w & 0x80) sp->regs[SP_STATUS_REG] &= ~SP_STATUS_INTR_BREAK;
-    if (w & 0x100) sp->regs[SP_STATUS_REG] |= SP_STATUS_INTR_BREAK;
+    if ((w & 0x180) == 0x80) sp->regs[SP_STATUS_REG] &= ~SP_STATUS_INTR_BREAK;
+    if ((w & 0x180) == 0x100) sp->regs[SP_STATUS_REG] |= SP_STATUS_INTR_BREAK;
 
     /* clear / set signal 0 */
-    if (w & 0x200) sp->regs[SP_STATUS_REG] &= ~SP_STATUS_SIG0;
-    if (w & 0x400) sp->regs[SP_STATUS_REG] |= SP_STATUS_SIG0;
+    if ((w & 0x600) == 0x200) sp->regs[SP_STATUS_REG] &= ~SP_STATUS_SIG0;
+    if ((w & 0x600) == 0x400)
```
Comment (GhostlyDark) [inline src/device/rcp/rsp/rsp_core.c:175]: It seems there's missing something here though. This is where code between upstream core and the one from parallel-n64 differ, as the latter follows up with multiline code that cannot be reused for upstream without further changes.

```
    {
         sp->regs[SP_STATUS_REG] |= SP_STATUS_SIG0;
         if (sp->audio_signal)
             signal_rcp_interrupt(sp->r4300, MI_INTR_SP);
     }
```

Maybe:
```
if ((w & 0x600) == 0x400) sp->regs[SP_STATUS_REG] |= SP_STATUS_SIG0;
```

File: src/device/rcp/rsp/rsp_core.c [hunk 2]
```diff
@@ -64,6 +64,10 @@ static void do_sp_dma(struct rsp_core* sp, const struct sp_dma* dma)
             post_framebuffer_write(&sp->dp->fb, dramaddr - length, length);
             dramaddr+=skip;
         }
+
+    sp->regs[SP_MEM_ADDR_REG] = memaddr & 0xfff;
```
Comment (loganmc10) [inline src/device/rcp/rsp/rsp_core.c:68]: these need to be indented

File: src/device/rcp/rsp/rsp_core.c [hunk 3]
```diff
@@ -77,6 +81,10 @@ static void do_sp_dma(struct rsp_core* sp, const struct sp_dma* dma)
             }
             dramaddr+=skip;
         }
+
+    sp->regs[SP_MEM_ADDR_REG] = memaddr & 0xfff;
```
Comment (loganmc10) [inline src/device/rcp/rsp/rsp_core.c:85]: these also need to be indented

File: src/device/rcp/rsp/rsp_core.c [hunk 4]
```diff
@@ -266,6 +276,7 @@ void write_rsp_regs(void* opaque, uint32_t address, uint32_t value, uint32_t mas
     {
     case SP_STATUS_REG:
         update_sp_status(sp, value & mask);
+        return;
```
Comment (loganmc10) [inline src/device/rcp/rsp/rsp_core.c:279]: this is redundant and doesn't need to be added

**Review by loganmc10 — CHANGES_REQUESTED** [review]

File: src/device/rcp/rsp/rsp_core.c (hunk 4 above)
Comment (mpharoah) [reply src/device/rcp/rsp/rsp_core.c]: Good for clarity though.

**Comment by Narann** [comment]
Thanks ! I also agree about putting those in a separated PR.

**Pull request merged**